### DISCUSSION TITLE
An auth example that uses directives.

### DISCUSSION THREAD

**Discussion by mikecann** [enisdenjo/graphql-ws#392] [Q&A] [answered]
### Story

As a server developer I would like to use auth directives on my schema to mark certain subscriptions as "User" or "Admin" only. Apollo gives a good tutorial on how to accomplish this: https://www.apollographql.com/docs/apollo-server/security/authentication/#with-custom-directives.

An example of this might look like:

```
type Subscription {
  battleUpdated(id: ID): Battle! #this subscription can be accessed by anonymous users
  battleCreated: Battle! @auth(requires: USER) #this one requires a logged in user to access
}
```

It is however not clear how to achieve this in graphql-ws.

So it would be useful to have an example to see how this could be done.

### Acceptance criteria

- server developer is able to quickly read the readme and see how to implement directive based authentication

**Comment by enisdenjo** [comment, marked as answer]
Auth directives is framework/implementor schema specific feature, not a part of `graphql-ws` which is just a transport library.

You can use the [`context` hook](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#context) to populate the GraphQL context with whatever you please and leverage the auth directives within your schema implementation. I also recommend taking a look at [`graphql-shield`](https://www.graphql-shield.com/).

**Reply by mikecann** [reply]
Apologies for the late reply, not sure how I missed this. Fantastic thanks, ill look into graphql-shield and perhaps ditch my directive based approach.
